src/main/java/com/mongodb/kafka/connect/sink/converter/LazyBsonDocument.java

```diff
@@ -127,10 +148,25 @@ public BsonDocument clone() {
 
   private BsonDocument getUnwrapped() {
     if (unwrapped == null) {
-      try {
-        unwrapped = supplier.get();
-      } catch (Exception e) {
-        throw new DataException("Unexpected data conversion exception.", e);
+      switch (dataType) {
+        case KEY:
+          try {
+            unwrapped = converter.apply(record.keySchema(), record.key());
+          } catch (Exception e) {
+            throw new DataException(
+                format("Could not convert key `%s` into a BsonDocument.", record.key()), e);
```

Comment:
By using a BiFunction the record can be accessed so meaningful error messages can be outputted.